### PR TITLE
[ci] Run unittest and compile examples on macOS CI

### PR DESCRIPTION
**Pull request by salkinium**
- Running unittests and compiling examples on macOS via TravisCI.
- Fixes(?) CMake scripts for toolchain detection(?) on macOS

cc @strongly-typed

### CODE REVIEW

**Comment by salkinium** [comment]
cc @strongly-typed You want to take this over? I'm too annoyed by Travis CI by this point.

**Comment by strongly-typed** [comment]
I'll try my best, I knew the pain of TravisCI beforehand ... 😂

**Comment by salkinium** [comment]
With your Makefile fix, only the zmqpp linkage errors remain, but I can reproduce them locally too, so they are not a CI issue, but maybe a Homebrew issue. I remember these examples compiling with an older version of the Formula.

**Comment by strongly-typed** [comment]
> With your Makefile fix, only the zmqpp linkage errors remain, but I can reproduce them locally too, so they are not a CI issue, but maybe a Homebrew issue. I remember these examples compiling with an older version of the Formula.

I was investigating the link errors as well. It links well with clang, but not with GNU g++-9.

**Comment by salkinium** [comment]
Should we just ignore it for now?

**Comment by salkinium** [comment]
@strongly-typed do you want to open a new PR with these changes and I'll close this one?

**Comment by salkinium** [comment]
Or maybe you can force push to my fork? Then the PR remains useful and the links valid.

**Comment by strongly-typed** [comment]
I'm currently working on the TravisCI. Once this is done, I'll push to your fork on this PR then.

**Comment by strongly-typed** [comment]
I am not allowed to push to your repo (anymore)?

TravisCI finished successfully the [Linux examples on macOS](https://travis-ci.com/github/strongly-typed/modm/builds/162445728#L3511).

I will add more tests on TravisCI later. For now, we gain some confidence, that modm can do something on macOS.

**Comment by salkinium** [comment]
I've added you as a contributer, you need to accept the invite tho.

File: tools/build_script_generator/cmake/cmake_scripts/configure-gcc.cmake.in [hunk 1]
```diff
@@ -40,38 +40,31 @@ else()
...
-    set(CMAKE_C_COMPILER "${TOOL_EXECUTABLE_PREFIX}gcc${TOOL_EXECUTABLE_SUFFIX}")
-    set(CMAKE_CXX_COMPILER "${TOOL_EXECUTABLE_PREFIX}g++${TOOL_EXECUTABLE_SUFFIX}")
-endif()
+set(CMAKE_TRY_COMPILE_TARGET_TYPE STATIC_LIBRARY)
+set(CMAKE_C_COMPILER "${TOOL_EXECUTABLE_PREFIX}gcc${TOOL_EXECUTABLE_SUFFIX}" CACHE INTERNAL "c")
+set(CMAKE_CXX_COMPILER "${TOOL_EXECUTABLE_PREFIX}g++${TOOL_EXECUTABLE_SUFFIX}" CACHE INTERNAL "c++")
 set(CMAKE_ASM_COMPILER "${CMAKE_C_COMPILER}"     CACHE INTERNAL "asm")
 
+set(CMAKE_AR           "${TOOL_EXECUTABLE_PREFIX}ar"      CACHE INTERNAL "ar")
 set(CMAKE_AR           "${TOOL_EXECUTABLE_PREFIX}ar"      CACHE INTERNAL "ar")
```
Comment (salkinium) [inline tools/build_script_generator/cmake/cmake_scripts/configure-gcc.cmake.in]: ?

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+  - brew update
+  - brew tap osx-cross/avr
+  - brew tap osx-cross/arm
+  - brew install scons git doxygen boost gcc avr-gcc arm-gcc-bin cmake || true
+  - brew upgrade boost gcc git || true
+  - export PATH=/Users/travis/Library/Python/3.7/bin:$PATH
+  - pip3 install --user modm
+
+script:
+  - (cd examples && ../tools/scripts/examples_compile.py linux)
```
Comment (salkinium) [inline .travis.yml:15]: Don't we also want to compile at least the hosted darwin unit tests and a few examples of AVR and STM32? Not all, it'll take too long.

Comment (strongly-typed) [reply .travis.yml:15]: On my host ...

```
➜  modm-str-ci_macos git:(feature/ci_macos) (cd test && make compile-arduino-nano)
lbuild -c config/arduino-nano.xml  -p ../build/generated-unittest/arduino-nano build --no-log
[WARNING] user.modm-test.test.platform.gpio: GPIO test has been hardcoded to a Nucleo-64 board. When porting make sure this test does not damage your board!

ERROR: In Module 'modm:platform:core':
In file '../src/modm/platform/core/avr/module.lb:65' in function 'build':
        env.template("delay_impl.hpp.in")

Error in template '../src/modm/platform/core/avr/delay_impl.hpp.in:47':
	}) : ({
%% if f_cpu <= 6e6
		constexpr int32_t cycles = (F_CPU / 1e6);
		for (int16_t _us = us; _us > 0; _us -= cycles)

TemplateSyntaxError: expected token 'end of statement block', got 'e6'
make: *** [compile-arduino-nano] Error 1
```

I enabled some, but we need to fix this first ...

Comment (salkinium) [reply .travis.yml:15]: Can't reproduce on my local machine and you seen to have fixed it anyways?

**Comment by salkinium** [comment]
~~The hosted unittests need to be compiled with gcc, however, macOS is weird and links gcc etc to clang. I prevent this by installing gcc via brew and using explicitly using gcc-9. The [SConscript explicitly searches for a gcc with a suffix](https://github.com/modm-io/modm/blob/develop/tools/build_script_generator/scons/resources/SConscript.in#L25), so I'm not sure why this is broken.~~ You used `make run-hosted-linux` instead of `make run-hosted-darwin`.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  - brew update
+  - brew tap osx-cross/avr
+  - brew tap osx-cross/arm
+  - brew install scons git doxygen boost gcc avr-gcc arm-gcc-bin cmake || true
+  - brew upgrade boost gcc git || true
+  - export PATH=/Users/travis/Library/Python/3.7/bin:$PATH
+  - pip3 install --user modm
+
+script:
+  - (cd test && make run-hosted-linux)
```
Comment (salkinium) [inline .travis.yml]: Oh, this is the issue.
```suggestion
  - (cd test && make run-hosted-darwin)
```

**Review by salkinium — COMMENTED** [review]
YES I CAN APPROVE MY OWN PULL REQUESTS, GITHUB!!!

**Comment by salkinium** [comment]
Ready to merge, @strongly-typed ?

**Comment by strongly-typed** [comment]
Totally! LGTM.

**Pull request merged**